util/list.c

```diff
@@ -0,0 +1,757 @@
+
+	list->nb_elements++;
+
+	return SUCCESS;
+}
+
+int32_t list_add_last(struct list_desc *list_desc, void *data)
+{
+	struct list_elem	*prev;
+	struct list_elem	*next;
+	struct list_elem	*elem;
+	struct _list_desc	*list;
+
+	if (!list_desc)
+		return FAILURE;
+	list = list_desc->priv_desc;
+
+	prev = list->last;
+	next = NULL;
+	elem = create_element(data, prev, next);
+	if (!elem)
+		return FAILURE;
+
+	update_links(prev, elem, next);
+
+	update_desc(list, list->first, elem);
+
+	list->nb_elements++;
+
+	return SUCCESS;
+}
+
+int32_t list_add_idx(struct list_desc *list_desc, void *data, uint32_t idx)
+{
+	struct _list_desc	*list;
+
+	if (!list_desc)
+		return FAILURE;
+	list = list_desc->priv_desc;
+
+	/* If there are no elements the creation of an iterator will fail */
+	if (list->nb_elements == 0 || idx == 0)
+		return list_add_first(list_desc, data);
+	if (list->nb_elements == idx)
+		return list_add_last(list_desc, data);
+
+	list->l_it.elem = list->first;
+	if (SUCCESS != iterator_move(&(list->l_it), idx))
+		return FAILURE;
+
+	return iterator_insert(&(list->l_it), data, 0);
+}
+
+int32_t list_add_find(struct list_desc *list_desc, void *data)
+{
+	struct list_elem	*elem;
+	struct _list_desc	*list;
+
+	if (!list_desc)
+		return FAILURE;
+	list = list_desc->priv_desc;
+
+
+	/* Based on place iterator */
+	elem = list->first;
+	while (elem) {
+		if (0 > list->comparator(data, elem->data))
+			break;
+		elem = elem->next;
+	}
+	if (elem == NULL) {
+		list->l_it.elem = list->last;
+		return iterator_insert(&(list->l_it), data, 1);
+	} else {
+		list->l_it.elem = elem;
+		return iterator_insert(&(list->l_it), data, 0);
+	}
+
+}
+
+int32_t list_edit_first(struct list_desc *list_desc, void *new_data)
+{
+	struct _list_desc	*list;
+
+	if (!list_desc)
+		return FAILURE;
+
+	list = list_desc->priv_desc;
+	list->first->data = new_data;
+
+	return SUCCESS;
+}
+
+int32_t list_edit_last(struct list_desc *list_desc, void *new_data)
+{
+	struct _list_desc	*list;
+
+	if (!list_desc)
+		return FAILURE;
+
+	list = list_desc->priv_desc;
+	list->last->data = new_data;
+
+	return SUCCESS;
+}
+
+int32_t list_edit_idx(struct list_desc *list_desc, void *new_data, uint32_t idx)
+{
+	struct _list_desc	*list;
+
+	if (!list_desc)
+		return FAILURE;
+	list = list_desc->priv_desc;
+
+	list->l_it.elem = list->first;
+	if (SUCCESS != iterator_move(&(list->l_it), idx))
+		return FAILURE;
+
+	return iterator_edit(&(list->l_it), new_data);
+}
+
+int32_t list_edit_find(struct list_desc *list_desc, void *new_data,
+		       void *cmp_data)
+{
+	struct _list_desc	*list;
+
+	if (!list_desc)
+		return FAILURE;
+	list = list_desc->priv_desc;
+
+	list->l_it.elem = list->first;
+	if (SUCCESS != iterator_find(&(list->l_it), cmp_data))
+		return FAILURE;
+
+	return iterator_edit(&(list->l_it), new_data);
+}
+
+void* list_read_first(struct list_desc *list_desc, int32_t *result)
+{
+	struct _list_desc	*list;
+
+	set_result(result, FAILURE);
+	if (!list_desc)
+		return FAILURE;
```

Comment:
Return integer type on a `void *` return. This ties to my earlier point that we would benefit by keeping consistency and returning a status and having output parameters to get data.

Reply:
I forgot to modify this when I changed to the approach of returning a void *